File: runner/client/src/mill/client/MillServerLauncher.java
package mill.client;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;
import mill.client.lock.Locks;
import mill.constants.Util;

/// Mill daemon specific {@link ServerLauncher}.
public abstract class MillServerLauncher extends ServerLauncher {
  final int serverInitWaitMillis = 10000;
  final Streams streams;
  final Map<String, String> env;
  final String[] args;

  /// For testing in memory, we need to pass in the locks separately, so that the
  /// locks can be shared between the different instances of `ServerLauncher` the
  /// same way file locks are shared between different Mill client/server processes
  final Optional<Locks> memoryLock;

  final int forceFailureForTestingMillisDelay;

  public MillServerLauncher(
    Streams streams,
    Map<String, String> env,
    String[] args,
    Optional<Locks> memoryLock,
    int forceFailureForTestingMillisDelay
  ) {
    this.streams = streams;
    this.env = env;
    this.args = args;
    this.memoryLock = memoryLock;
    this.forceFailureForTestingMillisDelay = forceFailureForTestingMillisDelay;
  }

  /**
   * Starts a Mill server
   *
   * @return the server process if available, or None
   */
  public abstract LaunchedServer initServer(Path daemonDir, Locks locks) throws Exception;

  public int run(Path daemonDir, String javaHome, Consumer<String> log) throws Exception {
    Files.createDirectories(daemonDir);

    var initData = new ClientInitData(
      /* interactive */ Util.hasConsole(),
      BuildInfo.millVersion,
      javaHome,
      args,
      env,
      ClientUtil.getUserSetProperties()
    );

    Locks locks;
    if (memoryLock.isPresent()) {
      locks = memoryLock.get();
    } else {
      locks = Locks.files(daemonDir.toString());
    }
    log.accept("launchOrConnectToServer: " + locks);

    try (
      var launched = launchOrConnectToServer(
        locks,
        daemonDir,
        serverInitWaitMillis,
        () -> initServer(daemonDir, locks),
        serverDied -> {
          System.err.println("Server died during startup:");
          System.err.println(serverDied.toString());
          System.exit(1);
        },
        log,
        true /*openSocket*/
      )
    ) {
      log.accept("runWithConnection: " + launched);
      var result = runWithConnection(
        launched.socket,
        streams,
        false,
        rawServerStdin -> {
          log.accept("Sending init data: " + initData);
          try {
            initData.write(rawServerStdin);
            log.accept("Init data sent.");
          } catch (IOException e) {
            throw new RuntimeException(e);
          }
        },
        () -> {
          log.accept("running client logic");
          forceTestFailure(daemonDir, log);
        },
        "MillServerLauncher[" + launched.socket.getLocalSocketAddress() + " -> "
          + launched.socket.getRemoteSocketAddress() + "]"
      );
      log.accept("runWithConnection exit code: " + result);
      return result;
    }
  }

  private void forceTestFailure(Path daemonDir, Consumer<String> log) {
    if (forceFailureForTestingMillisDelay > 0) {
      log.accept(
        "Force failure for testing in " + forceFailureForTestingMillisDelay + "ms: " + daemonDir);
      try {
        Thread.sleep(forceFailureForTestingMillisDelay);
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
      throw new RuntimeException("Force failure for testing: " + daemonDir);
    } else {
      log.accept("No force failure for testing: " + daemonDir);
    }
  }
}

File: runner/launcher/src/mill/launcher/MillServerLauncher.scala
package mill.launcher

import mill.api.daemon.SystemStreams
import mill.client.{ClientUtil, LaunchedServer, ServerLauncher}
import mill.constants.BuildInfo
import mill.constants.EnvVars
import mill.client.lock.Locks
import mill.constants.Util
import mill.util.Jvm
import mill.rpc.RpcConsole

import java.io.{BufferedReader, InputStreamReader, PrintStream}
import java.net.Socket
import java.util.concurrent.atomic.AtomicInteger

class MillServerLauncher(
    streamsOpt: Option[SystemStreams],
    env: Map[String, String],
    args: Seq[String],
    forceFailureForTestingMillisDelay: Int,
    useFileLocks: Boolean,
    initServerFactory: (os.Path, Locks) => LaunchedServer,
    millVersion: String = BuildInfo.millVersion,
    jvmOpts: Seq[String],
    millRepositories: Seq[String]
) {
  private val serverInitWaitMillis = 10000

  def run(daemonDir: os.Path, javaHome: Option[os.Path], log: String => Unit): Int = {
    os.makeDir.all(daemonDir)
    // `Locks.forDirectory` opens a real file lock; pass it the on-disk path, not the alias.
    val locks = Locks.forDirectory(Jvm.realAbs(daemonDir), useFileLocks)
    log(s"launchOrConnectToServer: $locks")

    val config = ServerLauncher.DaemonConfig(
      millVersion = millVersion,
      javaVersion = javaHome.map(_.toString).getOrElse(""),
      jvmOpts = jvmOpts,
      millRepositories = millRepositories,
      pathRelativizerBase = env.getOrElse(EnvVars.OS_LIB_PATH_RELATIVIZER_BASE, "")
    )

    val launched = ServerLauncher.launchOrConnectToServer(
      locks,
      daemonDir,
      serverInitWaitMillis,
      () => initServerFactory(daemonDir, locks),
      serverDied => throw new mill.api.MillException("Server died during startup: " + serverDied),
      s => log(s),
      true,
      config = config
    )

    try {
      log(s"runWithConnection (RPC): $launched")
      val result = runRpc(launched.socket.get, javaHome, daemonDir, log)
      log(s"runWithConnection exit code: $result")
      result
    } finally {
      // `launched.close()` closes the socket; `locks.close()` frees the file
      // channels/handles eagerly opened by `Locks.forDirectory`, which the lock
      // `release()`s do not. Guard each so a failure in one still attempts the other.
      try launched.close()
      catch { case _: Exception => }
      try locks.close()
      catch { case _: Exception => }
    }
  }

  private def runRpc(
      socket: Socket,
      javaHome: Option[os.Path],
      daemonDir: os.Path,
      log: String => Unit
  ): Int = {
    val stdout = streamsOpt.map(_.out).getOrElse(System.out)
    val stderr = streamsOpt.map(_.err).getOrElse(System.err)
    val exitCode = AtomicInteger(-1)
    try {
      val socketIn = BufferedReader(InputStreamReader(socket.getInputStream))
      val socketOut = PrintStream(socket.getOutputStream, true)

      val init = DaemonRpc.Initialize(
        interactive = Util.hasConsole(),
        clientPid = ProcessHandle.current().pid(),
        clientMillVersion = BuildInfo.millVersion,
        clientJavaVersion = javaHome.map(_.toString).getOrElse(""),
        clientJvmOpts = jvmOpts,
        args = args,
        env = env,
        userSpecifiedProperties = ClientUtil.getUserSetProperties(),
        millRepositories = millRepositories
      )

      val stdoutPs = PrintStream(stdout, true)
      val stderrPs = PrintStream(stderr, true)

      val stdoutHandler: RpcConsole.Message => Unit = {
        case RpcConsole.Message.Print(s) => stdoutPs.print(s)
        case RpcConsole.Message.Flush => stdoutPs.flush()
      }

      val stderrHandler: RpcConsole.Message => Unit = {
        case RpcConsole.Message.Print(s) => stderrPs.print(s)
        case RpcConsole.Message.Flush => stderrPs.flush()
      }

      val client = DaemonRpc.createClient(
        initialize = init,
        serverToClient = socketIn,
        clientToServer = socketOut,
        stdout = stdoutHandler,
        stderr = stderrHandler,
        runSubprocess = DaemonRpc.defaultRunSubprocessWithStreams(streamsOpt)
      )

      // For testing: run command in background while main thread throws after delay
      if (forceFailureForTestingMillisDelay > 0) {
        val commandThread = new Thread(() => {
          try {
            val result = client(DaemonRpc.ClientToServer.RunCommand())
            exitCode.set(result.exitCode)
          } catch { case _: Exception => }
        })
        commandThread.start()
        log(s"Force failure for testing in ${forceFailureForTestingMillisDelay}ms: $daemonDir")
        Thread.sleep(forceFailureForTestingMillisDelay)
        throw RuntimeException(s"Force failure for testing: $daemonDir")
      }

      val result = client(DaemonRpc.ClientToServer.RunCommand())
      exitCode.set(result.exitCode)
      client.close()
      exitCode.get()
    } catch {
      case e: RuntimeException if e.getMessage.startsWith("Force failure for testing:") =>
        throw e // Re-throw test exceptions
      case e: Exception =>
        e.printStackTrace(PrintStream(stderr))
        if (exitCode.get() < 0) exitCode.set(1)
        exitCode.get()
    }
  }
}
